### PR TITLE
Add tests and fixes for heap_free edge cases

### PR DESCRIPTION
**Pull request by rmn30**
heap_free and heap_claim: test, document and fix some edge cases.
  
  1) Use the capability base to look up the chunk to heap_free instead
  of the address. This guarantees the address is within the bounds of
  the original allocation, unlike the address. Same goes for heap_claim.
  
  2) Test that an attempt to free with a 'one-past-the-end' zero length
  capability fails.
  
  3) Test that an attempt to free fails if the bounds are not the full
  allocation.
  
  4) Test that it's OK to free a pointer to the interior of an
  allocation if it still has the full bounds.
  
  5) Don't allow claim or free of sealed capabilities: there's no reason
  to claim a sealed capability and for free the correct API is to use
  token_obj_destroy with the correct sealing key. Add the necessary
  checks and tests for this. Note that freeing fails for a different reason when
  using the unsealed handle to a sealed token: the unsealed handle does
  not encompass the whole allocation (including header) so the free
  fails with a different error code. Note that taking a claim to a
  suballocation and subsequently releasing it using heap_free is allowed.

### CODE REVIEW

File: tests/allocator-test.cc [hunk 1]
```diff
@@ -977,6 +977,22 @@ namespace
...
 
+		// It should not be possible to free sealed capabilities
+		TEST_EQUAL(heap_can_free(SECOND_HEAP, sealedPointer),
+		           -EINVAL,
+		           "heap_can_free should return error for sealed capability");
+		TEST_EQUAL(heap_free(SECOND_HEAP, sealedPointer),
+		           -EINVAL,
+		           "heap_free should return error for sealed capability");
+		// It should not be possible to free unsealed capabilities to sealed
+		// tokens (because the address doesn't include the header)
```
Comment (nwf) [inline tests/allocator-test.cc]: "the header" is maybe a bit too succinct, since there are several headers involved in allocation.  "the TokenObjectType header" perhaps?

**Review by nwf — APPROVED** [review]
One nit but LGTM

File: sdk/include/stdlib.h [hunk 2]
```diff
@@ -272,11 +272,17 @@ heap_claim_fast(Timeout         *timeout,
 }
 
 /**
- * Free a heap allocation.
+ * Free a heap allocation or release a claim made by `heap_claim`.
```
Comment (nwf) [inline sdk/include/stdlib.h]: :+1:

File: sdk/include/stdlib.h [hunk 3]
```diff
@@ -272,11 +272,17 @@ heap_claim_fast(Timeout         *timeout,
...
  *
- * Returns 0 on success, `-EINVAL` if `ptr` is not a valid pointer to the start
- * of a live heap allocation, or `-ENOTENOUGHSTACK` if the stack size is
- * insufficiently large to safely run the function.
+ * Returns:
+ *  - 0 on success
+ *  - `-EINVAL` if `heapCapability` is invalid or if  `ptr` is untagged, sealed,
+ *    is not a heap pointer or its address is outside its bounds.
```
Comment (nwf) [inline sdk/include/stdlib.h]:
```suggestion
 *  - `-EINVAL` if `heapCapability` is invalid or if  `ptr` is untagged, is sealed,
 *    is not a heap pointer, or has its address is outside its bounds.
```

Nit-picky.

File: tests/allocator-test.cc [hunk 4]
```diff
@@ -87,6 +87,65 @@ namespace
...
 	}
 
+	__noinline void test_heap_free()
+	{
+		debug_log("Testing heap_free edge cases");
+		Timeout          t{5};
+		Capability<void> p1 = heap_allocate(&t, MALLOC_CAPABILITY, 16);
+		TEST(p1.is_valid(), "Unable to make allocation for interior free test");
+
+		// Attempt to free with wrong heap capability
```
Comment (nwf) [inline tests/allocator-test.cc]: How had we not tested that already?

File: tests/allocator-test.cc [hunk 5]
```diff
@@ -87,6 +87,65 @@ namespace
...
+		  -EPERM,
+		  "attempt to free with wrong heap capability did not fail with "
+		  "-EPERM");
+		TEST(p1.is_valid_temporal(),
+		     "pointer invalid after failed free attempt with wrong heap");
+
+		// Attempt to free a pointer with (just) out of bounds address
+		Capability<void> p2 = p1;
+		p2.address()        = p2.top();
+		TEST(p2.is_valid(), "pointer invalid after setting address to top");
```
Comment (nwf) [inline tests/allocator-test.cc]: "This (`!p2.is_valid()`) can't happen because CHERI capabilities always guarantee one-past-the-end is tagged for C language support" or something like that?

File: tests/allocator-test.cc [hunk 6]
```diff
@@ -1005,6 +1066,11 @@ namespace
 		TEST(canFree != 0,
 		     "Should not be able to free a sealed capability with the wrong "
 		     "malloc capability but succeeded");
+		TEST_EQUAL(
+		  token_obj_destroy(
+		    MALLOC_CAPABILITY, sealingCapability, sealedPointer),
+		  -EPERM,
+		  "Freeing a sealed capability with the wrong allocator succeeded");
```
Comment (nwf) [inline tests/allocator-test.cc:1083]: "wrong quota"?

File: tests/allocator-test.cc (hunk 4 above)
Comment (rmn30) [reply tests/allocator-test.cc]: You mean the interior pointer free? It's possible we already do -- there are a lot of allocator tests! Actually I just noticed one of the tests I added is a duplicate (free with wrong quota) -- will remove that.

File: tests/allocator-test.cc (hunk 5 above)
Comment (rmn30) [reply tests/allocator-test.cc]: Added a comment.

File: tests/allocator-test.cc (hunk 6 above)
Comment (rmn30) [reply tests/allocator-test.cc:1083]: Removed this test as it's a duplicate.

File: sdk/core/allocator/main.cc [hunk 7]
```diff
@@ -832,13 +832,23 @@ namespace
 			return -EPERM;
 		}
 		Capability<void> mem{rawPointer};
-		if (!mem.is_valid())
+		/*
+		 * Validate the pointer. We do not permit freeing sealed pointers, or
+		 * pointers that are out of bounds. Note that address >= base is implied
+		 * by the combination of valid tag and CHERIoT capability encoding.
+		 */
+		if (!mem.is_valid() || mem.is_sealed() || mem.address() >= mem.top())
```
Comment (davidchisnall) [inline sdk/core/allocator/main.cc]: Rather than checking address < top, I think it would be cleaner above to do:

`mem.address() = mem.base();`

This ensures that any future use of address is sensible.

File: sdk/core/allocator/main.cc [hunk 8]
```diff
@@ -832,13 +832,23 @@ namespace
...
 		check_gm();
-		// Find the chunk that corresponds to this allocation.
-		auto *chunk = gm->allocation_start(mem.address());
+		/*
+		 * Find the chunk that corresponds to this allocation.  Use the base as
+		 * that cannot be further than one-past-the-end (because CHERI bounds
+		 * are monotone non-increasing), for us one-past-the-end is the next
+		 * chunk's header.
+		 */
+		auto *chunk = gm->allocation_start(mem.base());
```
Comment (davidchisnall) [inline sdk/core/allocator/main.cc:849]: With the above suggestions, this change is unnecessary.

File: sdk/core/allocator/main.cc (hunk 7 above)
Comment (rmn30) [reply sdk/core/allocator/main.cc]: That would prohibit freeing any interior pointer which is a change from the current behaviour although it is UB anyway so up to us. Freeing a claim on a sub allocation would still be possible provided `address==base`.

File: sdk/core/allocator/main.cc (hunk 8 above)
Comment (rmn30) [reply sdk/core/allocator/main.cc:849]: I agree that either `base` or `address` is fine here provided we have one of the checks on `address`. Note that `address < top` is actually slightly stricter than `base==address` because the latter allows a zero length capability to the header of the next allocation with `base==address==top`. As per @nwf 's comment that is OK.

File: sdk/core/allocator/main.cc (hunk 7 above)
Comment (rmn30) [reply sdk/core/allocator/main.cc]: OK, I misunderstood and thought you were suggesting to check that `mem.address == mem.base`. I think setting `mem.address` to `mem.base` is equivalent to using `mem.base` to look up the chunk below because the address of `mem` is not used otherwise. Setting the address to base would take an extra instruction vs. just calling `allocation_start` with the base so can't we just do that?
